controller.go

```diff
@@ -825,29 +825,41 @@ func parseAllocateConstraintsResponse(source interface{}, machine *machine) (Con
 	valid := coerced.(map[string]interface{})
 	constraintsMap := valid["constraints_by_type"].(map[string]interface{})
 	result := ConstraintMatches{
-		Interfaces: make(map[string]Interface),
-		Storage:    make(map[string]BlockDevice),
+		Interfaces: make(map[string][]Interface),
+		Storage:    make(map[string][]BlockDevice),
 	}
 
 	if interfaceMatches, found := constraintsMap["interfaces"]; found {
-		for label, value := range interfaceMatches.(map[string]interface{}) {
-			id := value.(int)
-			iface := machine.Interface(id)
-			if iface == nil {
-				return empty, NewDeserializationError("constraint match interface %q: %d does not match an interface for the machine", label, id)
+		matchMap := interfaceMatches.(map[string]interface{})
+		for label, values := range matchMap {
+			matches := values.([]interface{})
+			interfaces := make([]Interface, 0, len(matches))
+			for _, value := range matches {
+				id := value.(int)
```

Comment:
As discussed on IRC, I think it will be a lot simpler to parse the constraints map via an internal struct type (with possibly embedded types for storage and interfaces) with JSON tags on the fields.